### ISSUE TITLE
[Core] on_evicted callback

### ISSUE THREAD

**Issue by Atry** [ray-project/ray#45468]
### Description

A Ray API to register a callback triggered when a Ray object is evicted.

### Use case

Suppose I have a Ray actor that can create a Ray object that associates with some non-serializable states. In the following example, the non-serializable state is a temporary directory.

```python
class MyObject:
    pass

_non_serializable_states = {}

@ray.remote
class MyCreatorActor:
    def create(self):
        ref = ray.put(MyObject())
        _non_serializable_states[ref] = tempfile.mkdtemp()
        return ref
```

I want the following `on_evicted` function to be triggered on the creator actor's worker process when `ref` is evicted, i.e., when the Ray distributed reference counting for the Ray object is zero.

```python
def on_evicted(ref):
    shutil.rmtree(_non_serializable_states[ref])
    del _non_serializable_states[ref]
```

How can I modify `MyCreatorActor.create` to register `on_evicted`?

Is there a Ray API similar to `weakref.ref(obj, callback)` to register such a callback?

**Comment by jjyao** [comment]
@Atry We currently don't have it. I think you can do the application layer reference counting for your use case.

**Comment by Atry** [comment]
I think it is not possible to reliably do the application layer reference counting.

**Comment by jjyao** [comment]
Yes but if your application is simple, you can do application layer reference counting as a workaround for now to unblock you.
